modules/transformer-batch/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/transformer/batch/ShredJob.scala

```diff
@@ -151,13 +143,35 @@ class ShredJob[T](@transient val spark: SparkSession,
     // Final output
     transformer.sink(spark, config.output.compression, transformed, outFolder)
 
-    val batchTimestamps = timestampsAccumulator.value
+    val batchTimestamps = transformer.timestampsAccumulator.value
     val timestamps = Timestamps(jobStarted, Instant.now(), batchTimestamps.map(_.min), batchTimestamps.map(_.max))
 
     LoaderMessage.ShreddingComplete(outFolder, transformer.typesInfo, timestamps, config.output.compression, MessageProcessor, Some(LoaderMessage.Count(eventsCounter.value)))
   }
 }
 
+class TypeAccumJob(@transient val spark: SparkSession,
```

Comment:
We use this `TypeAccumJob` at the beginning of each run (early accumulation), cause when we transform event we have to know all types for given run, so we could provide field value if present and 'null' if there is no such type in event, correct? 
In 'Transformer' there is also `TypesAccumulator`, so it seems we accumulate all types twice?

So currently, we have:

1. Run job, detect all types from events.
2. Transformation - for every event traverse all the types, if there is a value for type - extract, if not - provide `null`.
3. Sink time - because of early (double?) accumulation we have all values and types, just build schema and we're done. 

How about a bit different approach:

1. No early accumulation, just usual one.
2. For every event - traverse **existing** types for event, extract them, no `null`s at that point. Just provide what is present.
3. Sink time - we have all types (accumulated as usual), at this point we can add missing nulls. We would have to detect types without values for each event and fill it with nulls.

I'm pointing it out, because I think in case of transformer-kinesis such early types accumulation is not possible, so it seems we would have to use 'add nulls later' approach in that case. It's just an idea and I'm not sure yet if it's even doable:) 
And if we are going to do this for transformer-kinesis we can have uniform approach and it also for batch transformer.

Reply:
The caveat is that it is possible to have different schema additions/revisions of same schema name. Let's say there are entities with following versions of same schema in the same batch: `1-0-1`, `1-0-2`, `1-1-0`, `1-2-0`. In this case, we want to create parquet schema with latest version which is `1-2-0` and we will create parquet data for the entities with other versions (1-0-1, 1-0-2, 1-1-0) with the parquet schema of `1-2-0`. 

Let's say we processed an event that contains some entity with schema version 1-0-1. Later on, we got an event that contains some entity with schema version 1-0-2. If we use the approach you proposed, we need to get back to old event and add nulls somehow. This might quickly get complicated.

Another problem is that there is no strict distinction between 'processing time' and 'sink time' in Spark. It is like pipeline. As events processed, they are written to destination. If we want to buffer them somehow to reprocess them again, we would need either memory as big as input data or we need to write it to disk which would be too slow.